File: 김현경/광물_캐기.py
def solution(picks, minerals):
    answer = 0
    # 곡괭이로 채굴 가능한 최대 광물 수
    minerals = minerals[:sum(picks) * 5]
    # 5개씩 묶어 리스트로 나눔
    new_minerals = [minerals[i:i+5] for i in range(0, len(minerals), 5)]
    
    # 광물에 대한 피로도 계산
    tireds = []
    for mineral in new_minerals:
        tired = [0, 0, 0]
        for m in mineral:
            if m == 'diamond':
                tired[0] += 1
                tired[1] += 5
                tired[2] += 25
            elif m == 'iron':
                tired[0] += 1
                tired[1] += 1
                tired[2] += 5
            else:
                tired[0] += 1
                tired[1] += 1
                tired[2] += 1
        tireds.append(tired)
        
    # 돌 곡괭이 사용했을 때의 피로도가 큰 순서대로 정렬
    # 돌 곡괭이로 채굴할 때 피로도가 많이 드는 광물부터 처리하기 위해서
    tireds = sorted(tireds, key=lambda x: -x[2]) 
    
    # 피로도가 높은 광물 그룹부터 곡괭이 사용
    # 다이아몬드 곡괭이가 남아 있으면 사용하고 아니면 넘어가고
    for tired in tireds:
        if picks[0] > 0:
            picks[0] -= 1
            answer += tired[0]
            continue
        elif picks[1] > 0:
            picks[1] -= 1
            answer += tired[1]
            continue
        elif picks[2] > 0:
            picks[2] -= 1
            answer += tired[2]
            continue
    return answer

File: 조성원/광물-캐기.js
const solution = (picks, minerals) => {
  let answer = 0;

  // 남아있는 곡괭이의 총 개수를 구하는 함수
  const getTotalPicks = () => picks.reduce((acc, cur) => acc + cur, 0);

  // 광물을 5개씩 묶어서 배열로 만든다.
  const chunks = [];
  for (let i = 0; i < minerals.length; i += 5)
    chunks.push(minerals.slice(i, i + 5));

  // 각 청크의 피로도를 계산
  chunks.forEach((chunk, index) => {
    const fatigue = chunk.reduce(
      (acc, cur) => acc + (cur === "diamond" ? 25 : cur === "iron" ? 5 : 1),
      0
    );

    chunks[index] = {
      chunk,
      fatigue,
    };
  });

  // 곡괭이가 부술 수 있는 청크의 수를 기준으로 내림차순 정렬
  chunks.splice(getTotalPicks());
  chunks.sort((a, b) => b.fatigue - a.fatigue);

  // 피로도 = 광물 경도 / 곡괭이 경도
  // 내림차순으로 정렬했으므로 다이아몬드 곡괭이부터 사용해서 광물을 캔다.
  while (getTotalPicks() > 0 && chunks.length > 0) {
    let hardness = 0;

    if (picks[0] > 0) {
      picks[0]--;
      hardness = 25;
    } else if (picks[1] > 0) {
      picks[1]--;
      hardness = 5;
    } else {
      picks[2]--;
      hardness = 1;
    }

    answer += chunks[0].chunk.reduce(
      (acc, cur) =>
        acc +
        (cur === "diamond"
          ? 25 / hardness
          : cur === "iron"
          ? Math.ceil(5 / hardness)
          : 1),
      0
    );

    chunks.shift();
  }

  return answer;
};
